src/sync/index.js

```diff
@@ -20,6 +23,139 @@ export type SyncDatabaseChangeSet = $Exact<{ [TableName<any>]: SyncTableChangeSe
+  // - What happens if I create then delete something locally - do i send it to sync?
+}
+
+export function applyRemoteChangesToCollection<T: Model>(
+  collection: Collection<T>,
+  changes: SyncTableChangeSet,
+): Promise<void> {
+  const { database } = collection
+  return database.action(async () => {
+    // - insert new records
+    //   - if already exists (error), update
+    // - destroy deleted records permanently
+    //   - if already deleted, ignore
+    //   - TODO: Should we delete with descendants? Or just let sync point to record to delete?
+    // - update records:
+    //   - if locally synced, update
+    //   - if locally updated (conflict):
+    //     - take changes from server, apply local changes from _changed, update
+    //   - if locally deleted:
+    //     - ignore (will push deletion later)
+    //   - if not found, insert
+    // TODO: Does the order between insert/update/destroy matter?
+
+    const { created, updated, deleted: deletedIds } = changes
+
+    const ids: RecordId[] = [...created, ...updated].map(({ id }) => id).concat(deletedIds)
+    const records = await collection.query(Q.where(columnName('id'), Q.oneOf(ids))).fetch()
+    const locallyDeletedIds = await database.adapter.getDeletedRecords(collection.table)
+
+    // Destroy records (if already marked as deleted, just destroy permanently)
+    const recordsToDestroy = records.filter(record => contains(record.id, deletedIds))
+    const deletedRecordsToDestroy = locallyDeletedIds.filter(id => contains(id, deletedIds))
+
+    await allPromises(record => record.destroyPermanently(), recordsToDestroy)
+    await database.adapter.destroyDeletedRecords(collection.table, deletedRecordsToDestroy)
+
+    // Insert and update records
+    const recordsToInsert = created.map(raw => {
+      const currentRecord = records.find(record => record.id === raw.id)
+      if (currentRecord) {
+        // TODO: log error -- record already exists, update instead
+        return prepareUpdateFromRaw(currentRecord, raw)
+      }
+
+      return prepareCreateFromRaw(collection, raw)
+    })
+
+    const recordsToUpdate = updated
+      .map(raw => {
+        const currentRecord = records.find(record => record.id === raw.id)
+
+        if (currentRecord) {
+          return prepareUpdateFromRaw(currentRecord, raw)
+        } else if (locallyDeletedIds.some(id => id === raw.id)) {
+          // Nothing to do, record was locally deleted, deletion will be pushed later
+        } else {
+          // Record doesn't exist (but should) — just create it
+          return prepareCreateFromRaw(collection, raw)
+        }
+      })
+      .filter(Boolean)
+
+    await database.batch(...recordsToInsert, ...recordsToUpdate)
+  })
+}
+
+export function applyRemoteChanges(
+  db: Database,
+  remoteChanges: SyncDatabaseChangeSet,
+): Promise<void> {
+  return db.action(async action => {
+    // TODO: Does the order of collections matter? Should they be done one by one? Or all at once?
```

Comment:
In case of this function that part doesn't matter since we're talking strictly local DB writes only. I decided it's best (for performance and consistency) to do all DB writes in a single big batch